src/include/storage/recovery/replication_log_provider.h

```diff
@@ -7,134 +7,195 @@
+  transaction::timestamp_t PopOAT() {
+    std::unique_lock lock(replication_latch_);
+    NOISEPAGE_ASSERT(!oats_.empty(), "Who's telling you to pop the OAT?");
+    transaction::timestamp_t oat = oats_.top().oat_;
+    oats_.pop();
+    replication_cv_.notify_all();
+    return oat;
+  }
 
-  // (Primary callback ID, buffers) that have arrived in packets from the master node
-  // TODO(Gus): For now, these buffers will most likely be allocated on demand by the network layer. We could consider
-  // having a buffer pool if allocation becomes a bottleneck
-  std::queue<std::pair<uint64_t, std::unique_ptr<network::ReadBuffer>>> arrived_buffer_queue_;
+ private:
+  /** @return True if left > right. False otherwise. */
+  static bool CompareBatches(const replication::RecordsBatchMsg &left, const replication::RecordsBatchMsg &right) {
+    return left.GetBatchId() > right.GetBatchId();
+  }
 
-  /**
-   * Callbacks that should be invoked on the primary once the current set of buffers that have been taken off the
-   * arrived buffer queue have been processed.
-   */
-  std::vector<uint64_t> primary_ackables_;
+  /** A pair of OAT and associated batch ID that must be processed before the OAT. */
+  struct OATPair {
+    transaction::timestamp_t oat_;             ///< The last transaction inclusive that is safe to be applied.
+    replication::record_batch_id_t batch_id_;  ///< The last batch inclusive that must be processed before applying OAT.
+  };
 
-  // Synchronisation primitives to synchronise arrival of packets and process termination
-  std::mutex replication_latch_;
-  std::condition_variable replication_cv_;
+  /** @return True if left > right. False otherwise. */
+  static bool CompareOATs(const OATPair &left, const OATPair &right) { return left.batch_id_ > right.batch_id_; }
+
+  /** @return True if the next batch has arrived. Assumes replication_latch_ is held. */
+  bool NextBatchReady() {
+    // If there are no batches, then the next batch certainly has not arrived.
+    if (received_batch_queue_.empty()) {
+      return false;
+    }
+    NOISEPAGE_ASSERT(received_batch_queue_.top().GetBatchId() > last_batch_popped_, "Duplicate batch added?");
+    // The next batch is ready for application if the batch ID is consecutive.
+    bool top_batch_is_next =
+        received_batch_queue_.top().GetBatchId() == replication::RecordsBatchMsg::NextBatchId(last_batch_popped_);
+    return top_batch_is_next;
+  }
+
+  // bounded set of 100 elements
 
   /**
-   * @return true if log file contains more records, false otherwise. Can be blocking if waiting for message from
-   * master, or master has timed out.
+   * @return        If replication has ended, false.
+   *                If there are more log records that can be immediately applied, true.
+   *                If there are no such log records and replication has not yet ended, blocks.
    */
   bool HasMoreRecords() override {
```

Comment:
I feel like there's probably a better name for this function. It doesn't really tell you if there or more records or not, but blocks the current until there are more records or replication has ended.